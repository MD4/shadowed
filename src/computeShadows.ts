import type { Bounds, Segment2, Vec2 } from './entities';
import { sortClockwise } from './sortClockwise';
import { isNotNull } from './isNotNull';
import { getWallProjectionToBounds } from './getWallProjectionToBounds';
import { isWallInBounds, padBounds } from './boundsUtils';

/**
 * @hidden
 */
const computeShadow = (bounds: Bounds, light: Vec2) => (
  wall: Segment2,
): Vec2[] | null => {
  const projection = getWallProjectionToBounds(bounds, light, wall);

  return projection.length >= 2 ? projection.concat(wall.a, wall.b) : null;
};

/**
 * Computes casted shadows by a light for a given set of walls within specified bounds.
 * @param bounds A rectangle where the shadows will be contained.
 * @param walls Segments that will cast shadows.
 * @param light Position of the light.
 *
 * @example
 * ```ts
 * computeShadows(
 *   {
 *     topLeft: { x: 0, y: 0 },
 *     bottomRight: { x: 640, y: 360 },
 *   },
 *   [{
 *     a: { x: 20, y: 80 },
 *     b: { x: 80, y: 30}
 *   }],
 *   { x: 100, y: 200 },
 * );
 * ```
 */
export const computeShadows = (
  bounds: Bounds,
  walls: Segment2[],
  light: Vec2,
): Vec2[][] =>
  walls
    .filter(isWallInBounds(bounds))
    .map(computeShadow(padBounds(bounds), light))
    .filter(isNotNull)
    .map(sortClockwise);
