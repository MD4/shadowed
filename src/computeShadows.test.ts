import { computeShadows } from './computeShadows';

describe('computeShadows', () => {
  it('compute shadows', () => {
    expect(
      computeShadows(
        {
          topLeft: { x: 0, y: 0 },
          bottomRight: { x: 640, y: 360 },
        },
        [
          // Polygon #1
          { a: { x: 100, y: 150 }, b: { x: 120, y: 50 } },
          { a: { x: 120, y: 50 }, b: { x: 200, y: 80 } },
          { a: { x: 200, y: 80 }, b: { x: 140, y: 210 } },
          { a: { x: 140, y: 210 }, b: { x: 100, y: 150 } },

          // Polygon #2
          { a: { x: 100, y: 200 }, b: { x: 120, y: 250 } },
          { a: { x: 120, y: 250 }, b: { x: 60, y: 300 } },
          { a: { x: 60, y: 300 }, b: { x: 100, y: 200 } },

          // Polygon #3
          { a: { x: 200, y: 260 }, b: { x: 220, y: 150 } },
          { a: { x: 220, y: 150 }, b: { x: 300, y: 200 } },
          { a: { x: 300, y: 200 }, b: { x: 350, y: 320 } },
          { a: { x: 350, y: 320 }, b: { x: 200, y: 260 } },

          // Polygon #4
          { a: { x: 340, y: 60 }, b: { x: 360, y: 40 } },
          { a: { x: 360, y: 40 }, b: { x: 370, y: 70 } },
          { a: { x: 370, y: 70 }, b: { x: 340, y: 60 } },

          // Polygon #5
          { a: { x: 450, y: 190 }, b: { x: 560, y: 170 } },
          { a: { x: 560, y: 170 }, b: { x: 540, y: 270 } },
          { a: { x: 540, y: 270 }, b: { x: 430, y: 290 } },
          { a: { x: 430, y: 290 }, b: { x: 450, y: 190 } },

          // Polygon #6
          { a: { x: 400, y: 95 }, b: { x: 580, y: 50 } },
          { a: { x: 580, y: 50 }, b: { x: 480, y: 150 } },
          { a: { x: 480, y: 150 }, b: { x: 400, y: 95 } },
        ],
        { x: 200, y: 200 },
      ),
    ).toMatchSnapshot();
  });
  it('excludes out-of-bounds walls', () => {
    expect(
      computeShadows(
        {
          topLeft: { x: 0, y: 0 },
          bottomRight: { x: 640, y: 360 },
        },
        [{ a: { x: 500, y: 400 }, b: { x: 700, y: 350 } }],
        { x: 200, y: 200 },
      ),
    ).toStrictEqual([]);
  });
});
