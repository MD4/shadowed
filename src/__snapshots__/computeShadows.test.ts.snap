// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeShadows compute shadows 1`] = `
Array [
  Array [
    Object {
      "x": 120,
      "y": 50,
    },
    Object {
      "x": 100,
      "y": 150,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 99.99749498737597,
    },
    Object {
      "x": -0.00001,
      "y": -0.00001,
    },
    Object {
      "x": 93.33589687531934,
      "y": -0.000009999999974752427,
    },
  ],
  Array [
    Object {
      "x": 200.00200000010074,
      "y": -0.000010000127304010675,
    },
    Object {
      "x": 200,
      "y": 80,
    },
    Object {
      "x": 120,
      "y": 50,
    },
    Object {
      "x": 93.33589687531934,
      "y": -0.000009999999974752427,
    },
  ],
  Array [
    Object {
      "x": 200.00200000010074,
      "y": -0.000010000127304010675,
    },
    Object {
      "x": 200,
      "y": 80,
    },
    Object {
      "x": 140,
      "y": 210,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 233.33127944776646,
    },
    Object {
      "x": -0.00001,
      "y": -0.00001,
    },
  ],
  Array [
    Object {
      "x": 140,
      "y": 210,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 233.33127944776646,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 99.99749498737597,
    },
    Object {
      "x": 100,
      "y": 150,
    },
  ],
  Array [
    Object {
      "x": 120,
      "y": 250,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 324.99722501723954,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 199.99799999989926,
    },
    Object {
      "x": 100,
      "y": 200,
    },
  ],
  Array [
    Object {
      "x": 120,
      "y": 250,
    },
    Object {
      "x": 60,
      "y": 300,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 342.854129613256,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 324.99722501723954,
    },
  ],
  Array [
    Object {
      "x": 100,
      "y": 200,
    },
    Object {
      "x": 60,
      "y": 300,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 342.854129613256,
    },
    Object {
      "x": -0.000010000000003174137,
      "y": 199.99799999989926,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": -0.00001,
    },
    Object {
      "x": 640.00001,
      "y": 360.00001,
    },
    Object {
      "x": 199.99839999989945,
      "y": 360.0000099964427,
    },
    Object {
      "x": 200,
      "y": 260,
    },
    Object {
      "x": 220,
      "y": 150,
    },
    Object {
      "x": 280.00232400939456,
      "y": -0.000010000000088439265,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": -0.00001,
    },
    Object {
      "x": 640.00001,
      "y": 200.0044000001016,
    },
    Object {
      "x": 300,
      "y": 200,
    },
    Object {
      "x": 220,
      "y": 150,
    },
    Object {
      "x": 280.00232400939456,
      "y": -0.000010000000088439265,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 200.0044000001016,
    },
    Object {
      "x": 640.00001,
      "y": 360.00001,
    },
    Object {
      "x": 399.9959125509912,
      "y": 360.00001000000003,
    },
    Object {
      "x": 350,
      "y": 320,
    },
    Object {
      "x": 300,
      "y": 200,
    },
  ],
  Array [
    Object {
      "x": 399.9959125509912,
      "y": 360.00001000000003,
    },
    Object {
      "x": 199.99839999989945,
      "y": 360.0000099964427,
    },
    Object {
      "x": 200,
      "y": 260,
    },
    Object {
      "x": 350,
      "y": 320,
    },
  ],
  Array [
    Object {
      "x": 400.00401004020125,
      "y": -0.000010000000003174137,
    },
    Object {
      "x": 400.00401004020125,
      "y": -0.000010000000003174137,
    },
    Object {
      "x": 340,
      "y": 60,
    },
    Object {
      "x": 360,
      "y": 40,
    },
  ],
  Array [
    Object {
      "x": 461.54389480487794,
      "y": -0.000009999999974752427,
    },
    Object {
      "x": 370,
      "y": 70,
    },
    Object {
      "x": 360,
      "y": 40,
    },
    Object {
      "x": 400.00401004020125,
      "y": -0.000010000000003174137,
    },
  ],
  Array [
    Object {
      "x": 461.54389480487794,
      "y": -0.000009999999974752427,
    },
    Object {
      "x": 370,
      "y": 70,
    },
    Object {
      "x": 340,
      "y": 60,
    },
    Object {
      "x": 400.00401004020125,
      "y": -0.000010000000003174137,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 163.33776305196326,
    },
    Object {
      "x": 640.00001,
      "y": 182.404406638336,
    },
    Object {
      "x": 450,
      "y": 190,
    },
    Object {
      "x": 560,
      "y": 170,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 163.33776305196326,
    },
    Object {
      "x": 640.00001,
      "y": 290.5928238676802,
    },
    Object {
      "x": 540,
      "y": 270,
    },
    Object {
      "x": 560,
      "y": 170,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 290.5928238676802,
    },
    Object {
      "x": 640.00001,
      "y": 360.00001,
    },
    Object {
      "x": 608.8768653689003,
      "y": 360.00001,
    },
    Object {
      "x": 430,
      "y": 290,
    },
    Object {
      "x": 540,
      "y": 270,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 182.404406638336,
    },
    Object {
      "x": 640.00001,
      "y": 360.00001,
    },
    Object {
      "x": 608.8768653689003,
      "y": 360.00001,
    },
    Object {
      "x": 430,
      "y": 290,
    },
    Object {
      "x": 450,
      "y": 190,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": -0.00001,
    },
    Object {
      "x": 640.00001,
      "y": 26.32087110192967,
    },
    Object {
      "x": 580,
      "y": 50,
    },
    Object {
      "x": 400,
      "y": 95,
    },
    Object {
      "x": 580.9616564125967,
      "y": -0.000010000000031595846,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": 26.32087110192967,
    },
    Object {
      "x": 640.00001,
      "y": 121.43310994097587,
    },
    Object {
      "x": 480,
      "y": 150,
    },
    Object {
      "x": 580,
      "y": 50,
    },
  ],
  Array [
    Object {
      "x": 640.00001,
      "y": -0.00001,
    },
    Object {
      "x": 640.00001,
      "y": 121.43310994097587,
    },
    Object {
      "x": 480,
      "y": 150,
    },
    Object {
      "x": 400,
      "y": 95,
    },
    Object {
      "x": 580.9616564125967,
      "y": -0.000010000000031595846,
    },
  ],
]
`;
